source/neuropod/bindings/java/src/main/java/com/uber/neuropod/NeuropodTensor.java

```diff
@@ -26,138 +26,192 @@
-      * @param index the index array
-      * @return the double element
-      */
-     double getDouble(long... index) {return 0.0;}
-
-     /**
-      * Gets a float element at the given index.
-      * <p>
-      * Can only be used when the tensor is FLOAT_TENSOR.
-      * For example, to get a float element from a 3rd-order tensor,
-      * the user should call getFloat(x, y, z)
-      *
-      * @param index the index array
-      * @return the float element
-      */
-     float getFloat(long... index) {return 0.0f;}
-
-     /**
-      * Gets a string element at the given index.
-      * <p>
-      * Can only be used when the tensor is STRING_TENSOR.
-      * For example, to get a string element from a 3rd-order tensor,
-      * the user should call getString(x, y, z)
-      *
-      * @param index the index array
-      * @return the string element
-      */
-     String getString(long... index) {return "";}
+    protected ByteBuffer buffer;
+
+    protected NeuropodTensor() {
+    }
+
+    /**
+     * Get the dims array which represnents the shape of a tensor.
+     *
+     * @return the shape array
+     */
+    public long[] getDims() {
+        return nativeGetDims(super.getNativeHandle());
+    }
+
+    /**
+     * Gets the number of elements of a tensor.
+     *
+     * @return the number of elements
+     */
+    public long getNumberOfElements() {
+        return nativeGetNumberOfElements(super.getNativeHandle());
+    }
+
+    /**
+     * Gets the type of a tensor
+     *
+     * @return the tensor type
+     */
+    public TensorType getTensorType() {
+        return nativeGetTensorType(super.getNativeHandle());
+    }
+
+    /**
+     * Flatten the tensor data and convert it to a long buffer.
+     * <p>
+     * Can only be used when the tensor is INT64_TENSOR. Will trigger
+     * a copy.
+     *
+     * @return the LongBuffer
+     */
+    public LongBuffer toLongBuffer() {
+        checkType(TensorType.INT64_TENSOR);
+        LongBuffer ret = LongBuffer.allocate((int) getNumberOfElements());
+        ret.put(buffer.asLongBuffer());
+        return ret;
+    }
+
+    /**
+     * Flatten the tensor data and convert it to a int buffer.
+     * <p>
+     * Can only be used when the tensor is INT32_TENSOR. Will trigger
+     * a copy.
+     *
+     * @return the IntBuffer
+     */
+    public IntBuffer toIntBuffer() {
+        checkType(TensorType.INT32_TENSOR);
+        IntBuffer ret = IntBuffer.allocate((int) getNumberOfElements());
+        ret.put(buffer.asIntBuffer());
+        return ret;
+    }
+
+    /**
+     * Flatten the tensor data and convert it to a float buffer.
+     * <p>
+     * Can only be used when the tensor is FLOAT_TENSOR. Will trigger
+     * a copy.
+     *
+     * @return the FloatBuffer
+     */
+    public FloatBuffer toFloatBuffer() {
+        checkType(TensorType.FLOAT_TENSOR);
+        FloatBuffer ret = FloatBuffer.allocate((int) getNumberOfElements());
+        ret.put(buffer.asFloatBuffer());
```

Comment:
Any reason we're making a copy and not just returning `ret.asFloatBuffer()`?

Similarly for the other `to*Buffer` methods

Reply:
If the tensor is created by the C++ side using infer method, the buffer will be only a wrapper for C++ memory(In the infer commit). If we do deallocate on the tensor, then the memory underlying the buffer will be corrupted. Thus it is unsafe if we expose the buffer to the user without a copy.

Reply:
Do you think it is a good idea to make tensor created in the Java side does not copy in this method, and tensor created in the C++ side copy in this method?